casper/src/main/scala/coop/rchain/casper/finality/Finalizer.scala

```diff
@@ -0,0 +1,192 @@
+                  }
+              }
+              // filter out empty results when no main parent and those out of scope
+              .map(_.collect { case (v, Some(meta)) if meta.blockNum > currLFBHeight => (v, meta) })
+
+            nextF.map(next => (out, next.nonEmpty.guard[Option].as(next)))
+          }
+          .evalMap(_.traverse {
+            // map visits to message agreements: validator v agrees on message m
+            case (v, m) =>
+              messageWeightMapF(m, dag).map { messageWeightMap =>
+                recordAgreement(messageWeightMap, v).map { agreement =>
+                  MessageAgreement(m, messageWeightMap, Map(agreement))
+                }
+              }
+          })
+          // collect successful agreements (by bonded validators as per message agreed on)
+          .map(_.collect { case Some(v) => v })
+          // flatten in a single stream
+          .map(Stream.emits(_))
+          .flatten
+      }
+
+    mkAgreementsStream.flatMap {
+      // while recording each agreement in agreements map
+      _.mapAccumulate(Map.empty[BlockMetadata, WeightMap]) {
+        case (acc, MessageAgreement(message, messageWeightMap, stakeAgreed)) =>
+          val curVal = acc.getOrElse(message, Map.empty)
+          require(
+            (stakeAgreed.keySet intersect curVal.keySet).isEmpty,
+            s"Logical error during finalization: message ${message.blockHash.show} got duplicate agreement."
+          )
+          val newVal = curVal ++ stakeAgreed // given requirement above its enough to just sum maps
+          (acc.updated(message, newVal), (message, messageWeightMap))
+      }
+      // output only target message of current agreement
+        .map {
+          case (fullAgreementsMap, (message, messageWeightMap)) =>
+            (message, messageWeightMap, fullAgreementsMap(message))
+        }
+        // filter only messages that cannot be orphaned
+        .filter {
+          case (_, messageWeightMap, agreeingWeightMap) =>
+            cannotBeOrphaned(messageWeightMap, agreeingWeightMap)
+        }
+        // compute fault tolerance
+        .evalMap {
+          case (message, messageWeightMap, agreeingWeightMap) =>
+            CliqueOracle
+              .computeOutput[F](
+                targetMsg = message.blockHash,
+                messageWeightMap = messageWeightMap,
+                agreeingWeightMap = agreeingWeightMap,
+                dag = dag
+              )
+              .map((message, _))
+        }
+        // first candidate that meets finalization criteria is new LFB
+        .filter { case (_, faultTolerance) => faultTolerance > faultToleranceThreshold }
+        .head
+        // execute all effects
+        .evalMap {
+          case (lfb, _) =>
+            newLfbEffect(lfb.blockHash).as(lfb.blockHash) <*
+              dag
+                .withAncestors(lfb.blockHash, dag.isFinalized(_).not)
+                .flatMap(_.toList.traverse(finalisationEffect))
```

Comment:
Because we don't have atomic operation on LMDB, here is possible for node to be killed between saving directly finalized block `newLfbEffect` and related operations in `finalisationEffect`.
Is it better to save LFB at the end? Can this be more safe so that `finalisationEffect` can be executed again?

Reply:
Makes sense, indeed, otherwise if node is killed `finalisationEffect ` might be never called for some messages that are ancestor of LFB. I'l push fix in commit addressing comments